pddl/formatter.py

```diff
@@ -40,16 +40,27 @@ def _print_predicates_with_types(predicates: Collection):
         else:
             result += f"({p.name}"
             for t in p.terms:
-                result += (
-                    f" ?{t.name} - {' '.join(t.type_tags)}"
-                    if t.type_tags
-                    else f" ?{t.name}"
-                )
+                if len(t.type_tags) > 1:
+                    result += f" ?{t.name} - (either {' '.join(t.type_tags)})"
+                else:
+                    result += (
+                        f" ?{t.name} - {list(t.type_tags)[0]}"
+                        if t.type_tags
+                        else f" ?{t.name}"
+                    )
             result += ") "
         result += " "
     return result.strip()
 
 
+def _print_types_with_parents(types: Dict):
+    result = ""
+    for t in sorted(types.keys()):
+        result += f"{t} - {types[t]}" if types[t] else f"{t}"
```

Comment:
Would this be capturing the ancestors? If it's just the direct types, wouldn't it need to be space-separated?  Maybe what's unclear here is the form of the `types` dict -- if it's mapping a type to it's direct parent (no `either` allowed), then maybe indicate that in a docstring or comment (the variable `types` can be ambiguous).

Reply:
Good point. I will make the form of the `types` dict more clear in the docstring.